cupy/core/fusion.pyx

```diff
@@ -1,887 +1,91 @@
 import functools
 import six
-import string
 
 import numpy
 
-import cupy
-from cupy.core._dtype import get_dtype
-from cupy.core import _errors
-from cupy.core import _kernel
-from cupy.core._kernel import _is_fusing
 from cupy.core import core
+from cupy.core import _fusion_analysis
+from cupy.core import _fusion_thread_local
+from cupy.core._fusion_shape import _AbstractDim
+from cupy.core._fusion_interface import _scalar  # NOQA
+from cupy.core._fusion_interface import _ndarray  # NOQA
+from cupy.core._dtype cimport get_dtype
 
 
-_thread_local = _kernel._thread_local
+_thread_local = _fusion_thread_local.thread_local
+_is_fusing = _fusion_thread_local.is_fusing
+_call_ufunc = _fusion_thread_local.call_ufunc
+_call_reduction = _fusion_thread_local.call_reduction
 
-cdef dict _kind_score = {
-    'b': 0,
-    'u': 1,
-    'i': 1,
-    'f': 2,
-    'c': 2,
-}
-
-cdef dict _dtype_to_ctype = {
-    numpy.dtype('float64'): 'double',
-    numpy.dtype('float32'): 'float',
-    numpy.dtype('float16'): 'float16',
-    numpy.dtype('complex128'): 'complex<double>',
-    numpy.dtype('complex64'): 'complex<float>',
-    numpy.dtype('int64'): 'long long',
-    numpy.dtype('int32'): 'int',
-    numpy.dtype('int16'): 'short',
-    numpy.dtype('int8'): 'signed char',
-    numpy.dtype('uint64'): 'unsigned long long',
-    numpy.dtype('uint32'): 'unsigned int',
-    numpy.dtype('uint16'): 'unsigned short',
-    numpy.dtype('uint8'): 'unsigned char',
-    numpy.dtype('bool'): 'bool',
-}
-
-cdef list _dtype_list = [numpy.dtype(_) for _ in '?bhilqBHILQefdFD']
 
 cdef tuple _acceptable_types = six.integer_types + (
```

Comment:
How about something like `_fusion_input_types` or `_fusion_argument_types`?
It's not clear who is going to accept them.